### ISSUE TITLE
Feature request

### ISSUE THREAD

**Issue by miku1958** [Leanny/leanny.github.io#163]
Is it possible to add a feature to "Gear Seed Recovery" to reverse the seed by clearing the skill and refreshing it with drink? 
For example, if I have three known skills and three refreshed skills, I can recover about 30 seeds, then I clear all the skills and refresh the fourth skill with a drink to get the real seeds.

I tried to write a APP myself but it looks like the algorithms are in gearSeed.wasm, and I'm not very familiar with disassembly or wasm....

**Comment by miku1958** [comment]
Yes, secondary search with possible seeds

To be more specific:
1. I first use the existing 4 skills to calculate 100 possible seeds
2. I add the 3 predicted skills to find 30 possible seeds, and filter out the seeds that don't match the first step, assuming there are 30 seeds left in the first step
3. Wash all the skills directly, then drink and go to the battle to get new skills, then see which one is compatible from the 30 seeds in the first step, one by one.

This is theoretically possible, right, but the workload is really too much for human.

For the page "technical write-up", I can't find the detail of "get_ability_weight", "get_branded_ability_with_drink" and "max_brand_num_drink"

**Comment by Leanny** [comment]
Thank you for the suggestion! I will add the refine button in a future update.

About the missing details: They are described in text form. Here is some pseudocode for them that can help you to understand what the methods might be doing. `Brand` is something generated from this here https://github.com/Leanny/leanny.github.io/blob/master/splat3/data/parameter/111/misc/spl__BrandTraitsParam.spl__BrandTraitsParam.json:
Drink is just the ability name:
```
uint32_t max_brand_num(Brand brand) {
    if(brand.isNeutral()) return 28;
    return 35;
}

uint32_t max_brand_num_drink(Brand brand, uint32_t drink) {
    if(brand.isNeutral()) return 26;
    if(brand.UnusualGearSkill == drink) return 34;
    if(brand.UsualGearSkill == drink) return 25;
    return 33;
}

uint32_t get_ability_weight(Brand brand, uint32_t ability_index) { // you can use this to generate the map that is used in the write-up. ability_index is defined by the ability order at the bottom of the write-up
    if(brand.UnusualGearSkill == ability_index) return 1;
    if(brand.UsualGearSkill == ability_index) return 10;
    return 2;
}

uint32_t get_branded_ability_with_drink(Brand brand, uint32_t ability_index, uint32_t drink) { // you can use this to generate the map that is used in the write-up. ability_index is defined by the ability order at the bottom of the write-up
    if(ability_index == drink) return 0;
    if(brand.UnusualGearSkill == ability_index) return 1;
    if(brand.UsualGearSkill == ability_index) return 10;
    return 2;
}

```

**Comment by Leanny** [comment]
Added with latest update. Thank you for the feedback!

**Comment by miku1958** [comment]
Thanks! Could you explain how to use the refine feature?

**Comment by miku1958** [comment]
Ok,let me reconfirm that refine will use the seeds in memory to count a new batch of seeds directly, and will this new batch of seeds be used to overwrite the seeds in memory, or will it just be used to filter the original batch of seeds?

**Comment by miku1958** [comment]
And I have a quick test
1. I first use the existing 4 skills to calculate 100 possible seeds
2. I add the 3 predicted skills, but the refine button is disable <- Not sure if I got it right

**Comment by Leanny** [comment]
it will overwrite the seeds in memory then.
For example:
![grafik](https://user-images.githubusercontent.com/1884062/194333023-efee1193-7458-49db-890e-2f05a20f962d.png)

Here I have 7 slots filled and calculated it once. The refine button is unlocked now, which means that I have less than 10k possible seeds.
Now I fill 8th slot and click "refine". It gives me a result immediately
![grafik](https://user-images.githubusercontent.com/1884062/194333393-a2e9189c-1618-4868-bd14-1cc81f92ec28.png)

**Comment by Leanny** [comment]
![grafik](https://user-images.githubusercontent.com/1884062/194379191-e2ed53d2-fc7b-46bb-9c25-838f2222c47e.png)
So like this? You get 1, 2, 3 via reroll, reset the game, use a drink and then get 4, 5, 6?

**Comment by miku1958** [comment]
<img width="838" alt="image" src="https://user-images.githubusercontent.com/24806909/194462333-f6d06aa7-be6f-44df-af99-66035da94309.png">

1. I have 4 existing  skills a.1, b.1 and c.1, this step can search for a large number of seeds, name it as s.1.
2. I use snail to predict the 3 new skills but not save the game, this step can create new seeds(s.2) from s.1, s.2 can be used to filter s.1, and get s.3 from s.1
   s.3 is a subset of s.1
   s.3 is not the subset of s.3
3. Wash all the skills directly, then drink and go to the battle to get new skills a.3 , then filter s.3 with a.3 to get a new s.4, s.4 is a subset of s.3
4. Repeat 3 and 2, until there is only one seed in s.x

By doing this, firstly I don't need to spend any 🐚, and secondly with any luck, the a.3 step will locate the seed

**Comment by miku1958** [comment]
> ```
> uint32_t max_brand_num(Brand brand) {
>     if(brand.isNeutral()) return 28;
>     return 35;
> }
> ```

What is Brand's data structure? Alias of UInt32?

**Comment by miku1958** [comment]
You are getting BrandData brand_data from get_ability_weight(brand) and getting ability_roll from brand_data[ability].weight, but get_ability_weight(brand) return a UInt32 variable?

**Comment by Leanny** [comment]
Ah, I understand now. I need to think about how to support this!

The code is just pseudo-code, so it is not working 100% and should only give you an idea about how things work.
`brand.isNeutral()` is `true` when brand has no usual/unusual ability (for example brand b99 - amiibo) 
![grafik](https://user-images.githubusercontent.com/1884062/194513863-9bf72e77-7a10-458d-8722-917c5c0a6f57.png)

Someone made a rust version for this algorithm, so maybe this can help you to understand it better https://github.com/Starwort/ability-miner/tree/master/src

**Comment by miku1958** [comment]
I still don't understand how can you get BrandData brand_data from get_ability_weight(brand) and getting ability_roll from brand_data[ability].weight, but get_ability_weight(brand) () return a UInt32 variable?

assume BrandData is a UInt32 Array and if get_ability_weight missing parameter here? The two pseudo-codes you gave are conflicting

**Comment by Leanny** [comment]
I have an update about your initial issue:
I added the option to define a `Refine Count` (the higher the count, the more results it stores). This allows the following in an early stage:

![grafik](https://user-images.githubusercontent.com/1884062/195361174-90f39bb5-b475-48df-acee-cfba97bc0dac.png)

When I do this, I get a certain number of possible seeds and "Refine Seed" is enabled. Now I can use drinks instead:
![grafik](https://user-images.githubusercontent.com/1884062/195361590-811a66ec-c439-4d6a-98d5-d13d941a67ba.png)

So basically I can remove the old slots and replace them with drink abilities. As long as I use "Refine Seed", I will only use the stores seeds.

Is this what you were looking for?

(This will be available with the next update)
